src/cpp/src/speculative_decoding/speculative_decoding_stateful_eagle3.cpp

```diff
@@ -0,0 +1,1242 @@
+    sequence_group->set_output_seq_len(sample_count);
+    sequence_group->set_num_validated_tokens(num_tokens_to_validate);
+
+    // Execute sampling
+    m_sampler.sample({sequence_group}, sliced_logits, is_validation_mode);
+    sequence_group->finish_iteration();
+
+    // Extract results based on mode
+    // TODO(top-k): For top-k, need to handle multiple sequences and their generated tokens
+    const auto& generated_ids = current_seq->get_generated_ids();
+    const size_t new_generated_len = generated_ids.size();
+
+    if (!is_validation_mode) {
+        OPENVINO_ASSERT(new_generated_len - prev_generated_len == sample_count,
+                        "Sampled token count mismatch: expected ",
+                        sample_count,
+                        ", got ",
+                        new_generated_len - prev_generated_len);
+
+        std::vector<int64_t> result_tokens(generated_ids.end() - sample_count, generated_ids.end());
+
+        log_debug(LogCategory::SAMPLE,
+                  "Sampled " + std::to_string(sample_count) + " token(s): " + format_tokens(result_tokens));
+
+        return (sample_count == 1) ? std::variant<int64_t, std::vector<int64_t>>(result_tokens[0]) : result_tokens;
+    } else {
+        // Validation mode: Sampler validates draft tokens and removes rejected ones
+        // Result includes accepted draft tokens + newly sampled token
+        const size_t result_count = new_generated_len - prev_generated_len + num_tokens_to_validate;
+        std::vector<int64_t> result_tokens(generated_ids.end() - result_count, generated_ids.end());
+
+        const size_t accepted_drafts = result_tokens.size() - 1;
+        log_debug(LogCategory::VALID,
+                  "Validation result: accepted " + std::to_string(accepted_drafts) + "/" +
+                      std::to_string(num_tokens_to_validate) + " drafts, tokens=" + format_tokens(result_tokens));
+
+        return result_tokens;
+    }
+}
+
+ov::Tensor Eagle3InferWrapperBase::get_logits() const {
+    return m_request.get_tensor("logits");
+}
+
+ov::Tensor Eagle3InferWrapperBase::get_hidden_features() const {
+    auto hidden_state = m_request.get_tensor("last_hidden_state");
+    const auto shape = hidden_state.get_shape();
+    OPENVINO_ASSERT(shape.size() == 3 && shape[0] == 1, "Invalid hidden state shape: ", format_shape(shape));
+
+    const size_t output_seq_len = shape[1];
+    const size_t hidden_size = shape[2];
+    const size_t actual_seq_len = m_request.get_tensor("input_ids").get_shape()[1];
+
+    if (output_seq_len == actual_seq_len)
+        return hidden_state;
+
+    OPENVINO_ASSERT(actual_seq_len <= output_seq_len,
+                    "Sequence length mismatch: actual=",
+                    actual_seq_len,
+                    ", output=",
+                    output_seq_len);
+    return ov::Tensor(hidden_state, {0, output_seq_len - actual_seq_len, 0}, {1, output_seq_len, hidden_size});
```

Comment:
It seems like `SLICE_OUT` actually should be done on `NPUW` side to `NPUW_LLM_MAX_GENERATION_TOKEN_LEN`...